src/shr_transport.h4

```diff
@@ -175,23 +175,36 @@ shmem_shr_transport_put(shmem_ctx_t ctx, void *target, const void *source,
 static inline void
 shmem_shr_transport_put_signal(shmem_ctx_t ctx, void *target,
                                const void *source, size_t len,
-                               uint64_t *sig_addr, uint64_t signal, int pe)
+                               uint64_t *sig_addr, uint64_t signal, int sig_op, int pe)
 {
 #if USE_MEMCPY
     memcpy(target, source, len);
-    *sig_addr = signal;
+    if (sig_op == SHMEM_SIGNAL_ADD) *sig_addr += signal;
+    else *sig_addr = signal;
 #elif USE_XPMEM
     shmem_transport_xpmem_put(target, source, len, pe,
                               shmem_internal_get_shr_rank(pe));
     shmem_internal_membar_acq_rel(); /* Memory fence to ensure target PE observes
                                         stores in the correct order */
+    uint64_t old_signal_val = 0;
+    if (sig_op == SHMEM_SIGNAL_ADD) {
+        shmem_transport_xpmem_get(&old_signal_val, sig_addr, sizeof(uint64_t), pe,
+                                  shmem_internal_get_shr_rank(pe));
+        signal += old_signal_val;
+    }
```

Comment:
This needs to be an atomic operation, use `shmem_shr_transport_atomic`.

Reply:
@jdinan `shmem_shr_transport_atomic` is enabled through `USE_SHR_ATOMICS`. Unless we use `--enable-shr-atomics` flag, we will not be able to use this API, right? Please let me know if I am missing something. 
The other question is, why does it need to be atomic? According to the spec, `All signaling operations on the signal data object completes as if performed atomically`. I thought the clause `as if performed atomically` gave the flexibility to use fi_inject or memcpy.

Reply:
If the target is waiting for several messages using the same signal variable, this will not work, right?

Reply:
Right, you would need to write code to support builds with and without shared memory atomics.

Reply:
Thanks @jdinan. One question, from the current code, there seems to be no relation between the flags `USE_SHR_ATOMICS` and `USE_XPMEM`/`USE_CMA`. Was there any particular reason behind this? I am thinking shouldn't `USE_SHR_ATOMICS` be defined when either of these on-node transports is defined?

Reply:
For nearly all builds, we can't enable shared memory atomics because the networking layer is not coherent with processor atomics. By default, we only enable shared memory copy to implement put/get. This is why shared memory atomics are a separate option, and disabled by default. CMA is essentially memory copy performed by the Linux kernel and it cannot support atomics.

Reply:
For CMA, we should directly use the transport layer atomics then, as you also suggested later. But, for XPMEM, can we say that enabling USE_XPMEM should enable USE_SHR_ATOMICS as well? We can use the shared memory atomics if `shmem_internal_get_shr_rank` returns something other than -1, right? But, currently, `shmem_shr_transport_use_atomic` also seems to be guarded by `USE_SHR_ATOMICS`.

Reply:
Correct, it is always safe to use shared memory put/get because OpenSHMEM says overlapping operations (e.g. between local shared memory and remote network transport accesses) leads to undefined behavior. Atomics do allow this overlapping. Therefore, it is only safe to use shared memory atomics when you know the transport layer is coherent with processor atomics. Because of this, enabling XPMEM only enables shared memory put/get and there is an additional flag to enable shared memory atomics.